### ISSUE TITLE
i have two gaussian_conditional ,when update model, why one update,another one dont update

### ISSUE THREAD

**Issue by battlepeace** [InterDigitalInc/CompressAI#217]
firstly,thank you for your work!
i have two gaussian_conditional ,when update model, why one update,another one dont update
self.gaussian_conditional = GaussianConditional(None)
self.gaussian_conditional_z = GaussianConditional(None)
![image](https://user-images.githubusercontent.com/49899631/224647627-9e3b5fba-0170-4a7a-872f-fdfd9180d381.png)
![image](https://user-images.githubusercontent.com/49899631/224647740-f2f29ae4-af34-4150-a3b4-b45aba88cf81.png)
![image](https://user-images.githubusercontent.com/49899631/224647766-9f73cbbb-ba21-4373-96b5-9a7ba8ea199b.png)

**Comment by YodaEmbedding** [comment]
Which version/commit of CompressAI is this? The latest version should handle all these things automatically.

https://github.com/InterDigitalInc/CompressAI/blob/master/compressai/models/base.py#L62

But if you want to figure out what's going on without `git pull`ing the latest changes:

- Add a `print("This function runs.")` statement when that function is called.
- Ensure the correct `load_state_dict` is called. `super()` sometimes may not refer to what you are [expecting](https://en.wikipedia.org/wiki/C3_linearization#Example).

**Comment by battlepeace** [comment]
my compressai version is 1.2.0,you say the new version can handle all these things automatically,I want to know if there are any other function updates in the new version?

**Comment by battlepeace** [comment]
I'm sorry I don't quite understand GaussianCondition .you means scale_table Is it a default table that won't change as the input image changes？

**Comment by YodaEmbedding** [comment]
Yes, their non-trainable parameters should all have the same default initialized values. This is fine.

The section I wrote about "Different scale tables" was just in case you wanted to customize it further.

**Comment by battlepeace** [comment]
Thank you very much for your answer. I have another question. There is a parameter in the code that is quality.For example, in the following model.
![image](https://user-images.githubusercontent.com/49899631/226092060-8aa4a213-1b69-4f5a-ad07-449b0dc512ce.png)
His quality parameters are as follows:
![image](https://user-images.githubusercontent.com/49899631/226092122-fb8edcc7-da47-427f-8e98-029939893b97.png)
I want to know what this parameter means. Is it N and M in the network model? Do you write this to facilitate selecting different N and M for experiments?

**Comment by YodaEmbedding** [comment]
Yes, these are `N, M`, and are used when calling the model's `__init__`:

https://github.com/InterDigitalInc/CompressAI/blob/ee91d536bd934fc1f8b1532f78db4c94072ae26d/compressai/models/google.py#L242

The quality parameter refers to different image quality targets that each separate model was trained on. This is done by changing the $lambda$ tradeoff hyperparameter in the loss, $L = R + \lambda D$. The lambdas used for each quality are listed here: https://interdigitalinc.github.io/CompressAI/zoo.html#training

**Comment by battlepeace** [comment]
I don't quite understand what you mean. What you mean is that the qualities of 1-8 are λ Are there different values for?

**Comment by YodaEmbedding** [comment]
The qualities are just a convenient way to number the different trained models which produce small files (quality=1) or large files (quality=8).

The lambdas shown were used to train these models. But you can train with any lambda you want.

**Comment by battlepeace** [comment]
oh,you mean that the specific training lambda is determined when I input training instructions, and the quality I choose will not affect the specific training process. For example, if I specify lambda=0.02, even if I choose quality=1, the lambda during training is still 0.02?

**Comment by YodaEmbedding** [comment]
Yes, that is correct.

**Comment by battlepeace** [comment]
ok，tooo much thanks to you

**Comment by battlepeace** [comment]
sorry,i have a another question.What exactly does aux_loss represent? When I train 10000 images with a batch size of 16 for 100 epochs, in the training output report, The value of aux_ loss is always high, ranging from around 15000 during initial training to around 5500 after 100 epochs. How can it be minimized?

**Comment by YodaEmbedding** [comment]
aux_loss is described here:

- https://interdigitalinc.github.io/CompressAI/models.html#compressai.models.CompressionModel.aux_loss
-

**Comment by battlepeace** [comment]
my train_one_epoch is same as original from examples/train.py.
what is s means?EntropyBottleneck() this function is entropy encoded for data, right? However, for example, the ScaleHyperprior model  from the h_a obtained  z ,z should be modeled using the Factorized Entropy Model. Has EntropyBottleneck() completed this operation?
i use mse to eval model,RD loss is mse loss?this is my output
[output.txt](https://github.com/InterDigitalInc/CompressAI/files/11010425/output.txt)

**Comment by YodaEmbedding** [comment]
The RD loss (0.6) looks good. My guess is that at least one of the entropy bottlenecks `.quantiles` are not being optimized.

EntropyBottleneck encodes each channel separately using some arbitrary distribution trained for each channel. That is indeed a factorized model, which is used on the z in ScaleHyperprior.

**Comment by battlepeace** [comment]
ok
[model.txt](https://github.com/InterDigitalInc/CompressAI/files/11014781/model.txt)

**Comment by YodaEmbedding** [comment]
Ah, I see, it's because of the custom `HighLowEntropyBottleneck`. Try this:

```python
ENTROPY_BOTTLENECKS = (EntropyBottleneck, HighLowEntropyBottleneck)

class my_model(JointAutoregressiveHierarchicalPriors):
    def load_state_dict(self, state_dict, strict=True):
        for name, module in self.named_modules():
            if not any(x.startswith(name) for x in state_dict.keys()):
                continue

            if isinstance(module, EntropyBottleneck):
                update_registered_buffers(
                    module,
                    name,
                    ["_quantized_cdf", "_offset", "_cdf_length"],
                    state_dict,
                )

            if isinstance(module, HighLowEntropyBottleneck):
                update_registered_buffers(
                    module,
                    name,
                    ["eb_l._quantized_cdf", "eb_l._offset", "eb_l._cdf_length",
                    "eb_h._quantized_cdf", "eb_h._offset", "eb_h._cdf_length"],
                    state_dict,
                )

            if isinstance(module, GaussianConditional):
                update_registered_buffers(
                    module,
                    name,
                    ["_quantized_cdf", "_offset", "_cdf_length", "scale_table"],
                    state_dict,
                )

        return nn.Module.load_state_dict(self, state_dict, strict=strict)

    def update(self, scale_table=None, force=False):
        if scale_table is None:
            scale_table = get_scale_table()
        updated = False
        for _, module in self.named_modules():
            if isinstance(module, ENTROPY_BOTTLENECKS):
                updated |= module.update(force=force)
            if isinstance(module, GaussianConditional):
                updated |= module.update_scale_table(scale_table, force=force)
        return updated

    def aux_loss(self) -> Tensor:
        loss = sum(m.loss() for m in self.modules() if isinstance(m, ENTROPY_BOTTLENECKS))
        return cast(Tensor, loss)
```

**Comment by battlepeace** [comment]
yes,it like this
class HighLowEntropyBottleneck(EntropyModel):


    def __init__(self, entropy_bottleneck_channels, alpha=0.5):
        super(HighLowEntropyBottleneck, self).__init__()
        beta = 1. - alpha
        eb_l_ch = int(alpha * entropy_bottleneck_channels)
        eb_h_ch = int(beta * entropy_bottleneck_channels)
        self.eb_l = EntropyBottleneck(channels=eb_l_ch)
        self.eb_h = EntropyBottleneck(channels=eb_h_ch)

    def forward(self, x):
        x_h, x_l = x
        x_q_h, x_h_likelihoods = self.eb_h(x_h)
        x_q_l, x_l_likelihoods = self.eb_l(x_l)
        return x_q_h, x_q_l, x_h_likelihoods, x_l_likelihoods

    def compress(self, x):
        x_h, x_l = x
        x_h_bits = self.eb_h.compress(x_h)
        x_l_bits = self.eb_l.compress(x_l)
        return x_h_bits, x_l_bits

    def decompress(self, strings, sizes):
        x_h_string = self.eb_h.decompress(strings[0], sizes[0])
        x_l_string = self.eb_l.decompress(strings[1], sizes[1])
        return x_h_string, x_l_string

    def update(self, force=False):
        return self.eb_l.update(force=force) & self.eb_h.update(force=force)

**Comment by YodaEmbedding** [comment]
The code you posted seems fine to me after a bit of review. I can't understand why it's not reducing the aux loss. The parameters for the aux optimizer are defined as follows:

```python
    parameters = {
        "net": {
            name
            for name, param in net.named_parameters()
            if param.requires_grad and not name.endswith(".quantiles")
        },
        "aux": {
            name
            for name, param in net.named_parameters()
            if param.requires_grad and name.endswith(".quantiles")
        },
    }
```

...however, since what you showed are the default `EntropyBottleneck` modules, they should have `.quantiles` parameters. The other modules don't have a `.quantiles` parameter, so no problems there, either...

**Comment by battlepeace** [comment]
your means is i debug it and watch the parameters of aux_optimizer and aux_optimizer.param_groups?

**Comment by YodaEmbedding** [comment]
Yes.

**Comment by battlepeace** [comment]
![image](https://user-images.githubusercontent.com/49899631/226281606-2e6c9bd7-2912-4c41-aef1-22d040abd945.png)
there may be some trouble？

**Comment by YodaEmbedding** [comment]
I think that looks fine for the untrained model. For the trained model, check:

```python
print(net.entropy_bottleneck.eb_l.quantiles)
print(net.entropy_bottleneck.eb_h.quantiles)
```

The values should all be different from `[-10, 0, 10]`, and look something similar to:

```python
>>> from compressai.zoo import mbt2018
>>> net = mbt2018(quality=2, pretrained=True)
>>> print(net.entropy_bottleneck.quantiles)
Parameter containing:
tensor([[[ 4.0214e-02,  4.3686e-02,  4.6902e-02]],

        [[-2.0358e+01,  6.2115e-01,  1.9330e+01]],

        [[ 5.0907e-02,  5.4218e-02,  5.7158e-02]],

        ...

        [[ 5.2013e-01,  5.3068e-01,  5.3920e-01]],

        [[ 1.6951e-01,  1.7440e-01,  1.7930e-01]],

        [[ 2.6147e-01,  2.6927e-01,  2.7660e-01]],

        [[ 2.6363e-01,  2.7121e-01,  2.7847e-01]]], requires_grad=True)
```

i.e. Most channels have a magnitude <1, but there's a couple that have a magnitude of roughly 20.

**Comment by battlepeace** [comment]
how i get it for the trained model?

**Comment by battlepeace** [comment]
i use update_model to run the checkpoint_best_loss.pth.tar and in this place i code it with    model_cls_or_entrypoint = models[architecture]
    if not isinstance(model_cls_or_entrypoint, type):
        model_cls = model_cls_or_entrypoint()
    else:
        model_cls = model_cls_or_entrypoint
    net = model_cls.from_state_dict(state_dict)
    print(net.entropy_bottleneck.eb_l.quantiles)
    print(net.entropy_bottleneck.eb_h.quantiles)

**Comment by battlepeace** [comment]
another question is i dont know why when i train it more than 100 epochs,the value of test MSE loss will be very unstable and large.you can comprare these two output.
[100 epoch output.txt](https://github.com/InterDigitalInc/CompressAI/files/11016893/100.epoch.output.txt)
[150epoch output.txt](https://github.com/InterDigitalInc/CompressAI/files/11016894/150epoch.output.txt)
first one is 100epoch training,test MSE loss will steadily decline
but second one is 150epoch training,test MSE loss acting very strange
Whether this is related to the fact that my test dataset has only 24 images, but I don't think it's the reason. It may be related to learning rate.
